day3/recipe/RecipesModel.h

```diff
@@ -0,0 +1,59 @@
+#ifndef RECIPESMODEL_H
+#define RECIPESMODEL_H
+
+#include <QtCore>
+#include <QtGui>
+#include <memory>
+
+
+class RecipesModel : public QAbstractListModel
+{
+    Q_OBJECT
+public:
+    enum DifficultyLevel {
+        LowDifficulty,
+        MidDifficulty,
+        HighDifficulty,
+        DifficultyCount,
+    };
+    Q_ENUM(DifficultyLevel)
+
+    static const QString difficultyStrings[];
+
+    enum RoleNames {
+        DescriptionRole = Qt::UserRole,
+        ImageSourceRole = Qt::UserRole + 1,
+        DifficultyRole = Qt::UserRole + 2,
+        PreparationTimeRole = Qt::UserRole + 3,
+        DifficultyStringRole = Qt::UserRole + 4,
+        PreferredRole = Qt::UserRole + 5
+    };
+
+    struct RecipeData {
+        RecipeData(QString descr, QString  imageSo, DifficultyLevel difficul, int prepTime, bool pref): description(descr), imageSource(imageSo), difficulty(difficul)
```

Comment:
In generale i parametri read-only di una funzione dovrebbero essere `const&`.
Dal C++11 in avanti però, se passi una variabile non-POD (es. `QString`) per valore allora puoi usare `std::move()` per inizializzare i membri, così da toglierti una copia. Quindi questo costruttore dovrebbe assumere una delle due forme:
```c++
RecipeData(const QString &descr, /*...*/) : description(descr), /*...*/ {}
```
opppure
```c++
RecipeData(QString descr /*...*/) : description(std::move(descr)) /*...*/ {}
```

Reply:
ok, usata varsione `const QString &`